lib/filewriter.js

```diff
@@ -0,0 +1,182 @@
+/**
+ * FileWriter -- fast non-streaming file appender for accepting buffered data
+ *
+ * Locks the file when appending to be safe with multiple writers.
+ * Also, it reopens the file periodically, to make it work with logrotate.
+ *
+ * Copyright (C) 2014-2015 Andras Radics
+ * Licensed under the Apache License, Version 2.0
+ */
+
+var fs = require('fs');
+var fse = require('./fs-ext.js');
+
+var flock = fse.flock;
+var flockSync = function( fd, mode ) {
+    try { fse.flockSync(fd, mode); return true }
+    catch (err) { return false }
+}
+
+fse.SEEK_SET = 0;
+fse.SEEK_CUR = 1;
+fse.SEEK_END = 2;
+
+module.exports = FileWriter;
+
+function FileWriter( filename, openmode ) {
+    'use strict';
+
+    if (!(this instanceof FileWriter)) return new FileWriter(filename, openmode);
+    if (!filename) throw new Error("missing filename");
+
+    this._isFileWriter = true;
+    this.filename = filename;
+    this.openmode = openmode || "a";
+    // do not overwrite contents when reopening file, use a instead of w
+    this.reopenmode = this.openmode[0] !== 'w' ? openmode : ((openmode[1] === '+' || openmode[1] === 'r') ? 'a+' : 'a');
```

Comment:
This line does too much and is a bit difficult to parse.  Can you separate it out more?